### DISCUSSION TITLE
CI/CD Strategy

### DISCUSSION THREAD

**Discussion by bennibbelink** [cyclus/cyclus#1593] [Ideas]
## Testing Goals
Every time changes are made to `main` branch: 
1. Confirm that Cyclus builds
2. Confirm that unit tests pass
3. Confirm that new Cyclus build does not break dependants
4. Push new packages to registry

Every time changes are made to repositories dependent on Cyclus (Cycamore, Cymetric, ..., I'll refer to them generically as ABC):
1. Confirm that ABC builds
2. Confirm that ABC unit tests pass
3. Push new packages to registry

### Maintain similar workflows in each repository
Currently (#1588) Cyclus has GitHub workflows in place that build and test Cyclus via Docker builds, and pushes the docker images to ghcr.io.
One solution is to incorporate more stages in the Docker build that tests compatibility with Cycamore.

Pros: easy to implement, extending the Cyclus workflow should be relatively easy

Cons: in order to meet the second set of testing goals we would need to copy the workflows (or a subset of them) to each dependent repo, this would likely require manually going through each repo to make common changes.  I could see this getting sloppy over time

### Use `repository_dispatch` to propagate workflows
GitHub supports a workflow trigger called [repository_dispatch](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#repository_dispatch).  Typically this is used for clients outside of the repo to trigger workflows via GitHub's REST API (docs for this particular [endpoint](https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#create-a-repository-dispatch-event)).  We can use [this](https://github.com/marketplace/actions/repository-dispatch) action in GitHub marketplace to trigger workflows between separate repositories (took a brief look at the repo, it essentially just constructs and executes the API call for you). 

Using this approach we could accomplish boths sets of testing goals and minimize redundancy across codebases.
#### On changes to Cyclus:
1. Perform tests relevant to ONLY cyclus
2. Dispatch an event to other repositories that trigger their respective testing workflows that could:
    - Pull and test using the latest version of cyclus from a registry OR
    - Use a version of Cyclus for testing as specified by the payload from Cyclus' dispatch
3. On successful completion of ABC workflows, push new package to registry
    - One nuance I want to point out is: Assume that changes are made to cyclus and it passes all tests defined in the cyclus workflows, then an event is sent out that triggers ABC workflows.  If for some reason ABC breaks, we probably don't want to push the cyclus image to a registry (even though it passed its own tests).  If we only had a single ABC repo (Cycamore let's say), then when the Cycamore workflow completes successfully it could dispatch an event back to Cyclus to trigger a workflow that pushes to a registry.  However, if we have more than one ABC this won't work.  A possible solution to this is chaining workflows together(?), i.e. Cyclus triggers ABC which triggers DEF which triggers ... XYZ which ultimately triggers Cyclus to push to a registry
#### On changes to ABC:
1. Simply execute a workflow that pulls the latest Cyclus release from a registry to use for tests

Repository dispatching also supports custom event types and payloads, so we can expand on it in the future to only trigger certain workflows, pass data from repo to repo, etc.

**Comment by gonuke** [comment]
Another option may be to maintain our own actions to be used across multiple repositories

**Reply by bennibbelink** [reply]
Hmm.. I think I might like this idea.  If all the repos pull from the same set of actions then we can make changes to our testing workflows without modifying each repo independently.  Also - if they all pull from the same TS module (that we write) we probably don't need to trigger workflows in other repos.

What did you have in mind?

**Comment by gonuke** [comment]
For reference, I'm leaving a link to this snapshot where we had a robust testing strategy in place using CircleCI:
* [cyclus](https://github.com/cyclus/cyclus/tree/c1910b904f22264285fc95506f994604ede22bb8)
* [cycamore](https://github.com/cyclus/cycamore/tree/f0971466f67d742ae60591d0c201038934948016)
* [cymetric](https://github.com/cyclus/cymetric/tree/acc20f5c807612b89a0077d2ce467dde8e5c035d)

It may not longer be functional at this point, but it does document how we chose to do it in the past.

**Reply by bennibbelink** [reply]
Looks like the cyclus workflow has jobs defined for testing cycamore and cymetric, but doesn't actually include them in the workflows (I'm guessing they're left over from an older testing strategy?). Instead it triggers the cymetric and cycamore repos to execute their own workflows. 

Also I'm not seeing where cyclus unit tests are being run in the cyclus workflow.  The workflow just deploys new images and triggers the other repos (please double check this, I'm not super familiar with CircleCI)

**Comment by bennibbelink** [comment]
What are our priorities for compatibility between repos?  If changes to Cyclus are made and they break Cycamore, do we want to accept or reject the cyclus changes?  Asking for clarification because if Cyclus updates supersede compatibility with the other repos then our CI/CD would be simpler

**Reply by gonuke** [reply]
This definitely the key question.  

My gut response is that we want to ensure a full working stack at release time, but each individual PR in cyclus is allowed to break downstream packages.  That said, we probably want advisory testing on each PR - it's good to know we broke something even if it doesn't prevent a PR from merging.  Perhaps such testing happens at time of merge of the PR and not at time of submission.

The resulting question is what this strategy means for changes in cycamore.  Cycamore should be forced to work with some version of cyclus, but not sure joe to specify this...

**Reply by bennibbelink** [reply]
Perhaps make the workflows depend on versioning?  We could maintain major version numbers across repos, so that if we release Cycamore v10.x.x it must work with Cyclus v10.y.y.  

I'm assuming any major changes to break things would start with Cyclus, so a new major Cyclus release v11.0.0 release doesn't need to be compatible with Cycamore since Cycamore v11.0.0 doesn't exist yet.  BUT if cycamore v11.0.0 exists, then Cyclus 11.0.1 must be compatible with Cycamore v11

Should be relatively easy to have workflows check the package registry for available versions and pull/test accordingly

**Comment by gonuke** [comment]
@bennibbelink and I discussed this in person and came up with the following testing strategy.

Let's define `KR` to refer to the last release of the Cyclus kernel, and `KM` to refer to the current head of the `main` branch, and `KC` refers to a currently checked out branch of the Cyclus kernel (e.g. for a PR).

Let's also define 'CR` to refer to the last release of the Cycamore package, and `CM` to refer to the `main` branch of Cycamore, and `CC` to refer to a current checked out branch of Cycamore (e.g. for a PR).

For each PR into `cyclus/main` we will test `KC` with `CR` and `CM`, with the following truth table for outcomes:
* `CR` passes, `CM` passes: everything is fine - allow merge to proceed
*  `CR` fails, `CM` passes: this may be an improvement being implemented in lock-step in both Cyclus & Cycamore, and already merged in Cycamore.  If so, allow merge to proceed.
*  `CR` fails, `CM` fails:  probably indicates that this PR is an improvement in Cyclus that will need a corresponding change in Cycamore - allow merge to proceed if a corresponding PR exists in Cycamore.
*  `CR` passes, `CM` fails:  probably indicates an improvement already merged in Cycamore that needs to be reflected in this kernel PR, rebase or update PR - confirm kernel update and rebase before merge

For each PR into `cycamore/main` we will test `CC` agianst `KR` and `KM`, with the following truth table for outcomes:
* `KR` passes , `KM` passes: everything is fine - allow merge to proceed
* `KR` fails, `KM` passes: this may be an improvement being implemented in lock-step in both Cyclus & Cycamore, and already merged in Cyclus - allow merge to proceed
* `KR` fails, `KM` fails: probably indicates that this PR is an improvement in Cycamore that needs a corresponding change in Cyclus - delay merge until the kernel is updated
* `KR` passes, `KM` fails: suggests that kernel update has occurred that needs matching Cycamore update - confirm cycamore update and rebase before merge
